README
## HEI
___
### Overview
___
The goal of hei is to calculate Healthy Eating Index scores. The Healthy Eating Index is a dietary metric designed by the USDA and NCI to gauge adherence to the US Dietary Guidelines.
### Installation
___
```
devtools::install_github("vpnagraj/hei")
```
### Getting Started
___
>`library(hei)`

hei can be used to perform two key calculations.

>`leg_all()` works on a dataset containing the necessary food patterns information for a collection of individuals (such as that recorded in NHANES and the FPED), returning a dataset containing additional variables representing protein categories with legumes appropriately allocated.

>`hei()` works on a dataset containing food patterns information that includes protein categories with legumes appropriately allocated (such as that returned by `leg_all()`) returning a dataset with scores for each of the twelve HEI components as well as a total score.

hei also includes `get_diet()` and `get_demo()` for retrieving NHANES dietary and demographic data from the web for a given year, respectively, as well as `combo()`, which will join an FPED data set with a dietary and a demographic dataset such as these.
### Related Work
___
hei utilizes several tidyverse packages including dplyr and readr. hei also uses the nhanesA package.

* [Read more about the tidyverse approach to data analysis.](https://github.com/tidyverse)

* [Learn about the NHANES survey package.](https://github.com/cjendres1/nhanes)

NEWS.md
# hei

## 0.1.0

- initial release on CRAN: <https://CRAN.R-project.org/package=hei>

## 0.2.0

- now including citation and usage message when package is attached
- bug fix for tests

tests/testthat.R
library(testthat)
library(hei)

test_check("hei")


tests/testthat/test_day.R
context("days")

test_that("day choice is enforced", {

    expect_error(get_diet(day = "1999th"))
    expect_error(get_fped(day = "1999th"))

})

tests/testthat/test_hei.R
context("HEI calculation")

test_that("HEI score is same as original validation data", {

    diet <- get_diet("2009/2010", "both")
    fped <- get_fped("2009/2010", "both")
    demog <- get_demo("2009/2010")

    heires <- hei(fped,diet,demog)

    expect_equal_to_reference(heires$HEI, file = "rds/valid.rds")

})

test_that("HEI verbose argument returns all columns", {

    diet <- get_diet("2009/2010", "both")
    fped <- get_fped("2009/2010", "both")
    demog <- get_demo("2009/2010")

    heires <- hei(fped,diet, demog, verbose = TRUE)

    expect_equal(ncol(heires), 70)

})

R/combo.R
#' Data set combination and manipulation for use in HEI scoring.
#'
#' \code{combo} returns a single data set combining the FPED and NHANES dietary and demographic data sets. Additionally, it allows for specification of an age threshold such that records with age values below this number will be excluded from the output. Furthermore, any rows for which dietary recall was deemed unreliable (DRSTZ != 1) are automatically dropped. Finally, additional necessary HEI inputs are generated from the existing columns, see "Value" section below.
#'
#' @param fped food pattern equivalent database data
#' @param diet dietary data from NHANES database
#' @param demograph demographic data from NHANES database
#' @param agethresh numeric threshold for age of survey participants to be included
#' @return Object of class \code{data.frame} representing a data set fully prepped for HEI score calculation. All columns from the three input data sets are preserved while a handful of additional derivative columns are included, specifically:
#' \itemize{
#' \item WHOLEFRT: T_F_CITMLB (citrus, melons, and berries) + T_F_OTHER (fruits excluding citrus, melons, and berries)
#' \item MONOPOLY: TMFAT (total monounsaturated fatty acids) + TPFAT (total polyunsaturated fatty acids)
#' \item ALLMEAT: T_PF_MPS_TOTAL (meat, poultry, seafood, organ meat, and cured meat) + T_PF_EGGS (eggs and egg substitutes) + T_PF_NUTSDS (peanuts, tree nuts, and seeds) + T_PF_SOY (soy products)
#' \item SEAPLANT: T_PF_SEAFD_HI (Seafood high in n-3 fatty acids) + T_PF_SEAFD_LOW (Seafood low in n-3 fatty acids) + T_PF_NUTSDS (peanuts, tree nuts, and seeds) + T_PF_SOY + (soy products)
#' \item ADDSUGC: T_ADD_SUGARS (added sugars) * 16
#' \item SOLFATC: T_SOLID_FATS (fats naturally present in meat, etc.) * 9
#' \item MAXALCGR: TKCAL (total calories) / 1000 * 13
#' \item EXALCCAL (amount of alcohol above acceptable threshold): (TALCO (total alcohol) - MAXALCGR) * 7
#' \item EMPTYCAL10: ADDSUGC + SOLFATC + EXALCCAL
#' }

combo <- function(fped, diet, demograph, agethresh = 2) {

    dat <- merge(fped, diet, all.x = TRUE, by = "SEQN")
    dat <- merge(dat, demograph, all.x = TRUE, by = "SEQN")
    dat <- subset(dat, dat$RIDAGEYR >= agethresh)
    dat <- subset(dat, dat$DRSTZ == 1)

    dat$WHOLEFRT <- dat$T_F_CITMLB + dat$T_F_OTHER
    dat$MONOPOLY <- dat$TMFAT + dat$TPFAT
    dat$ALLMEAT <- dat$T_PF_MPS_TOTAL + dat$T_PF_EGGS + dat$T_PF_NUTSDS + dat$T_PF_SOY
    dat$SEAPLANT <- dat$T_PF_SEAFD_HI + dat$T_PF_SEAFD_LOW + dat$T_PF_NUTSDS + dat$T_PF_SOY
    dat$ADDSUGC <- 16*dat$T_ADD_SUGARS
    dat$SOLFATC <- 9*dat$T_SOLID_FATS
    dat$MAXALCGR <- 13 *(dat$TKCAL/1000)
    dat$EXALCCAL <- ifelse(dat$TALCO <= dat$MAXALCGR, 0, 7*(dat$TALCO-dat$MAXALCGR))
    dat$EMPTYCAL10 <- dat$ADDSUGC + dat$SOLFATC + dat$EXALCCAL

    dat

}

R/leg_all.R
#' allocate legumes for HEI scoring
#'
#' @param dat data to be processed
#' @return data frame of dietary values with legumes allocated appropriately

leg_all <- function(dat) {

    dat$mbmax <- 2.5*(dat$TKCAL/1000)

    dat$meatleg <- ifelse(dat$ALLMEAT < dat$mbmax,
                          dat$T_V_LEGUMES*4,
                          0)

    dat$needmeat <- ifelse(dat$ALLMEAT < dat$mbmax,
                           dat$mbmax - dat$ALLMEAT,
                           0)

    dat$lallmeat <- ifelse(dat$meatleg <= dat$needmeat,
                           dat$ALLMEAT + dat$meatleg,
                           0)

    dat$lseaplant <- ifelse(dat$meatleg <= dat$needmeat,
                            dat$SEAPLANT + dat$meatleg,
                            0)

    dat$lvtotal <- ifelse(dat$meatleg <= dat$needmeat,
                          dat$T_V_TOTAL,
                          0)

    dat$lbeangrn <- ifelse(dat$meatleg <= dat$needmeat,
                           dat$T_V_DRKGR,
                           0)

    dat$extrameat <- ifelse(dat$meatleg > dat$needmeat,
                            dat$meatleg - dat$needmeat,
                            0)

    dat$extraleg <- ifelse(dat$meatleg > dat$needmeat,
                           dat$extrameat/4,
                           0)

    dat$lallmeat <- ifelse(dat$meatleg > dat$needmeat,
                           dat$ALLMEAT + dat$needmeat,
                           dat$lallmeat)

    dat$lseaplant <- ifelse(dat$meatleg > dat$needmeat,
                            dat$SEAPLANT + dat$needmeat,
                            dat$lseaplant)

    dat$lvtotal <- ifelse(dat$meatleg > dat$needmeat,
                          dat$T_V_TOTAL + dat$extraleg,
                          dat$lvtotal)

    dat$lbeangrn <- ifelse(dat$meatleg > dat$needmeat,
                           dat$T_V_DRKGR + dat$extraleg,
                           dat$lbeangrn)

    dat$lallmeat <- ifelse(dat$ALLMEAT >= dat$mbmax,
                           dat$ALLMEAT,
                           dat$lallmeat)

    dat$lseaplant <- ifelse(dat$ALLMEAT >= dat$mbmax,
                            dat$SEAPLANT,
                            dat$lseaplant)

    dat$lvtotal <- ifelse(dat$ALLMEAT >= dat$mbmax,
                          dat$T_V_TOTAL + dat$T_V_LEGUMES,
                          dat$lvtotal)

    dat$lbeangrn <- ifelse(dat$ALLMEAT >= dat$mbmax,
                           dat$T_V_DRKGR + dat$T_V_LEGUMES,
                           dat$lbeangrn)

    dat

}

R/get_demo.R
#' NHANES demographic data retrieval.
#'
#' \code{get_demo} returns an abridged National Health and Nutrition Examination Survey (NHANES) demographic data set for a given year.
#'
#' @param year year combination of data set to retrieve ... must be one of the following:
#' \itemize{
#' \item "2005/2006"
#' \item "2007/2008"
#' \item "2009/2010"
#' \item "2011/2012"
#' \item "2013/2014"
#' }
#' @return Object of class \code{data.frame} containing a modified NHANES demographic data set for a specified year. The complete NHANES demographic data set contains several more variables than are necessary for computing HEI scores. In fact, the only variable contained in the returned data set (in addition to an NHANES unique sequence identifier for each participant) is not, strictly speaking, even required itself, but becomes relevant if data sets from multiple separate NHANES iterations are concatenated, specifically:
#' \itemize{
#' \item SDDSRVYR: This variable represents the two-year data release cycle number (e.g. a value of “6” denotes NHANES 2009–2010).
#' }
#' @references \url{https://www.cdc.gov/nchs/nhanes/nhanes_questionnaires.htm}
#' @export
#' @examples
#' \dontrun{
#' get_demo("2009/2010")
#' }

get_demo <- function(year) {

    yearchoices <- c("D" = "2005/2006",
                     "E" = "2007/2008",
                     "F" = "2009/2010",
                     "G" = "2011/2012",
                     "H" = "2013/2014")

    try(if(!year %in% yearchoices)
        stop("must use valid year choice, see ?get_demo for valid choices",
             call. = FALSE))

    dbname <- paste0("DEMO_", names(which(yearchoices==year)))

    dat <- nhanesA::nhanes(dbname)

    keepers <- c("SEQN", "SDDSRVYR")
    dat <- dat[,names(dat) %in% keepers]

    # have to convert each column to numeric to remove labelling from NHANES
    dat <- data.frame(apply(dat, 2, as.numeric))

    dat

}

R/get_diet.R
#' NHANES dietary data retrieval.
#'
#' \code{get_diet} returns an abridged National Health and Nutrition Examination Survey (NHANES) dietary data set for a given year.
#'
#' @param year year combination of data set to retrieve ... must be one of the following:
#' \itemize{
#' \item "2005/2006"
#' \item "2007/2008"
#' \item "2009/2010"
#' \item "2011/2012"
#' \item "2013/2014"
#' }
#' @param day identifier for survey day to retrieve ... must be one of the following:
#' \itemize{
#' \item "first" (data for study participants from first day)
#' \item "second" (study participant data from second day)
#' \item "both" (data for study participants present in both days, with values averaged across the two days)
#' }
#' @return Object of class \code{data.frame} containing a modified NHANES dietary data set for a specified year and day. The complete NHANES dietary data set contains several more variables than are necessary for computing HEI scores. The data set returned by this function has been trimmed down to contain (in addition to an NHANES unique sequence identifier for each participant) only the columns of data needed for HEI score calculation, specifically:
#' \itemize{
#' \item TKCAL: Energy (kcal)
#' \item TSFAT: Total saturated fatty acids (gm)
#' \item TMFAT: Total monounsaturated fatty acids (gm)
#' \item TPFAT: Total polyunsaturated fatty acids (gm)
#' \item TSODI: Sodium (mg)
#' \item TALCO: Alcohol (gm)
#' }
#'
#' @references \url{https://www.cdc.gov/nchs/nhanes/nhanes_questionnaires.htm}
#' @export
#' @examples
#' \dontrun{
#' get_diet("2009/2010", "both")
#' get_diet("2009/2010", "first")
#' }

get_diet <- function(year, day) {

    yearchoices <- c("D" = "2005/2006",
                     "E"= "2007/2008",
                     "F"="2009/2010",
                     "G"= "2011/2012",
                     "H" = "2013/2014")

    try(if(!year %in% yearchoices)
        stop("must use valid year choice, see ?get_diet for valid choices",
             call. = FALSE))

    daychoices <- c("DR1"="first", "DR2"="second","both")

    try(if(!day %in% daychoices)
        stop("must use valid day choice, see ?get_diet for valid choices",
             call. = FALSE))

    if(day != "both") {

        dbname <- paste0(names(which(daychoices==day)),
                         "TOT_",
                         names(which(yearchoices==year)))

        dat <- nhanesA::nhanes(dbname)

        names(dat) <- gsub("DR[1-9]", "", names(dat))

        keepers <- c("SEQN",
                     "TKCAL",
                     "TSFAT",
                     "TALCO",
                     "TSODI",
                     "TMFAT",
                     "TPFAT")

        dat <- dat[,names(dat) %in% keepers]

        # have to convert each column to numeric to remove labelling from NHANES
        dat <- data.frame(apply(dat, 2, as.numeric))

    } else {

        dbname1 <- paste0(names(daychoices[1]),
                          "TOT_",
                          names(which(yearchoices==year))
                          )

        dat1 <- nhanesA::nhanes(dbname1)

        names(dat1) <- gsub("DR[1-9]", "", names(dat1))

        keepers <- c("SEQN",
                     "TKCAL",
                     "TSFAT",
                     "TALCO",
                     "TSODI",
                     "TMFAT",
                     "TPFAT")

        dat1 <- dat1[,names(dat1) %in% keepers]

        dbname2 <- paste0(names(daychoices[2]),
                          "TOT_",
                          names(which(yearchoices==year))
                          )

        dat2 <- nhanesA::nhanes(dbname2)

        names(dat2) <- gsub("DR[1-9]", "", names(dat2))

        keepers <- c("SEQN", "TKCAL", "TSFAT", "TALCO", "TSODI", "TMFAT", "TPFAT")

        dat2 <- dat2[,names(dat2) %in% keepers]

        dat <- rbind(dat1,dat2)

        # have to convert each column to numeric to remove labelling from NHANES
        dat <- data.frame(apply(dat, 2, as.numeric))

        dat <- dat[!is.na(dat$TSFAT),]

        dat <- stats::aggregate(. ~ SEQN, data = dat, FUN = "mean")

    }

    dat

}

R/get_fped.R
#' FPED data retrieval.
#'
#' \code{get_fped} returns an abridged Food Patterns Equivalents Database (FPED) data set for a given year.
#'
#' @param year year combination of data set to retrieve ... must be one of the following:
#' \itemize{
#' \item "2005/2006"
#' \item "2007/2008"
#' \item "2009/2010"
#' \item "2011/2012"
#' \item "2013/2014"
#' }
#' @param day identifier for survey day to retrieve ... must be one of the following:
#' \itemize{
#' \item "first" (data for study participants from first day)
#' \item "second" (study participant data from second day)
#' \item "both" (data for study participants present in both days, with values averaged across the two days)
#' }
#' @return Object of class \code{data.frame} containing a modified FPED data set for specified year and day. FPED proper contains several more variables than are necessary for computing HEI scores. The data set returned by this function has been trimmed down to contain (in addition to an NHANES unique sequence identifier for each participant, age, and dietary recall status) only the columns of data needed for HEI score calculation, specifically:
#' \itemize{
#' \item T_F_CITMLB: Intact fruits (whole or cut) of citrus, melons, and berries (cup eq.)
#' \item T_PF_EGGS: Eggs (chicken, duck, goose, quail) and egg substitutes (oz. eq.)
#' \item T_F_OTHER: Intact fruits (whole or cut); excluding citrus, melons, and berries (cup eq.)
#' \item T_PF_MPS_TOTAL: Total of meat, poultry, seafood, organ meat, and cured meat (oz. eq.)
#' \item T_PF_SOY: Soy products, excluding calcium fortified soy milk (soymilk) and mature soybeans (oz. eq.)
#' \item T_PF_SEAFD_HI: Seafood (finfish, shellfish, and other seafood) high in n-3 fatty acids (oz. eq.)
#' \item T_PF_SEAFD_LOW: Seafood (finfish, shellfish, and other seafood) low in n-3 fatty acids (oz. eq.)
#' \item T_PF_NUTSDS: Peanuts, tree nuts, and seeds; excludes coconut (oz. eq.)
#' \item T_SOLID_FATS: Fats naturally present in meat, poultry, eggs, and dairy (lard, tallow, and butter); hydrogenated or partially hydrogenated oils; shortening; palm, palm kernel and coconut oils; fat naturally present in coconut meat and cocoa butter; and 50\% of fat present in stick and tub margarines and margarine spreads (grams)
#' \item T_ADD_SUGARS: Foods defined as added sugars (tsp. eq.)
#' \item T_V_TOTAL: Total dark green, red and orange, starchy, and other vegetables; excludes legumes (cup eq.)
#' \item T_V_LEGUMES: Total dark green, red and orange, starchy, and other vegetables; excludes legumes (cup eq.)
#' \item T_V_DRKGR: Dark green vegetables (cup eq.)
#' \item T_F_TOTAL: Total intact fruits (whole or cut) and fruit juices (cup eq.)
#' \item T_G_WHOLE: Grains defined as whole grains and contain the entire grain kernel - the bran, germ, and endosperm (oz. eq.)
#' \item T_D_TOTAL: Total milk, yogurt, cheese, and whey. For some foods, the total dairy values could be higher than the sum of D_MILK, D_YOGURT, and D_CHEESE because miscellaneous dairy component composed of whey is not included in FPED as a separate variable. (cup eq.)
#' \item T_G_REFINED: Refined grains that do not contain all of the components of the entire grain kernel (oz. eq.)
#' }
#' @references \url{https://www.ars.usda.gov/northeast-area/beltsville-md/beltsville-human-nutrition-research-center/food-surveys-research-group/docs/fped-overview/}
#' @export
#' @examples
#' get_fped("2009/2010", "both")
#' get_fped("2005/2006", "first")

get_fped <- function(year, day) {

    yearchoices <- c("fped0506" = "2005/2006",
                     "fped0708"= "2007/2008",
                     "fped0910"="2009/2010",
                     "fped1112"= "2011/2012",
                     "fped1314" = "2013/2014")

    daychoices <- c("both" = "both",
                    "day1" = "first",
                    "day2" = "second")

    try(if(!year %in% yearchoices)
        stop("must use valid year choice, see ?get_fped for valid choices",
             call. = FALSE))

    try(if(!day %in% daychoices)
        stop("must use valid day choice, see ?get_fped for valid choices",
             call. = FALSE))

    fped <- paste0(names(which(yearchoices==year)),
                   "_",
                   names(which(daychoices==day))
    )

    eval(parse(text = fped))

}

R/hei.R
#' Individual HEI scores calculation.
#'
#' \code{hei} calculates a Health Eating Index (HEI) score for individuals in the National Health and Nutrition Examination Survey (NHANES) studies based on input dietary, demographic and Food Pattern Equivalent Database (FPED) data.
#'
#' @param fped Food Pattern Equivalent Database data; see \link{get_fped}
#' @param diet dietary data from NHANES database; see \link{get_diet}
#' @param demograph demographic data from NHANES database; see \link{get_demo}
#' @param agethresh numeric threshold for age in years of survey participants to be included; any individual less than the value specified will be excluded; defaults to 2
#' @param verbose Boolean indicating whether or not all columns from processed data should be output; default is \code{FALSE} designating only the following are included in the returned \code{data.frame}:
#' \itemize{
#' \item SEQN: Respondent sequence number
#' \item RIDAGEYR: Best age in years of the sample person at time of HH screening. Individuals 85 and over are topcoded at 85 years of age
#' \item HEI: Overall Health Eating Index score for the given participant
#' }
#' @return Object of class \code{data.frame}; defaults to only include columns for respondent identifier, age, and overall HEI score of each individual; this can be overridden with the \code{verbose} parameter to output all columns of the input data sets as well as 33 columns of calculated data related to HEI scoring and, significantly, a 70th column containing the total HEI score for each participant.
#' @export
#'
#' @references \url{https://www.cnpp.usda.gov/healthyeatingindex}
#'
#' @examples
#' \dontrun{
#' fped0910 <- get_fped("2009/2010", "both")
#' diet0910 <- get_diet("2009/2010", "both")
#' demo0910 <- get_demo("2009/2010")
#' hei(fped0910,diet0910,demo0910)
#'
#' fped0506 <- get_fped("2005/2006", "first")
#' diet0506 <- get_diet("2005/2006", "first")
#' demo0506 <- get_demo("2005/2006")
#' hei(fped0506,diet0506,demo0506, agethresh = 18)
#' }

hei <- function(fped, diet, demograph, agethresh = 2, verbose = FALSE) {

    dat <- combo(fped, diet, demograph, agethresh)

    dat <- leg_all(dat)

    # heiveg
    # total veggies
    dat$vegden <- dat$lvtotal / (dat$TKCAL/1000)
    dat$heiveg <- 5*(dat$vegden/1.1)
    dat$heiveg <- ifelse(dat$heiveg > 5, 5, dat$heiveg)

    # heibngrn
    # beans and greens
    dat$bngrden <- dat$lbeangrn / (dat$TKCAL/1000)
    dat$heibngrn <- 5*(dat$bngrden/0.2)
    dat$heibngrn  <- ifelse(dat$heibngrn  > 5, 5, dat$heibngrn)

    # heitotfr
    # total fruit
    dat$frtden <- dat$T_F_TOTAL/(dat$TKCAL/1000)
    dat$heitotfrt <- 5*(dat$frtden/0.8)
    dat$heitotfrt <- ifelse(dat$heitotfrt > 5, 5, dat$heitotfrt)

    # heiwholefrt
    # whole fruit
    dat$wholefrtden <- dat$WHOLEFRT/(dat$TKCAL/1000)
    dat$heiwholefrt <- 5*(dat$wholefrtden/0.4)
    dat$heiwholefrt <- ifelse(dat$heiwholefrt > 5, 5, dat$heiwholefrt)

    # heiwholegrain
    # whole grain
    dat$wholegrainden <- dat$T_G_WHOLE/(dat$TKCAL/1000)
    dat$heiwholegrain <- 10*(dat$wholegrainden/1.5)
    dat$heiwholegrain <- ifelse(dat$heiwholegrain > 10, 10, dat$heiwholegrain)

    # heidairy
    # dairy
    dat$dairyden <- dat$T_D_TOTAL/(dat$TKCAL/1000)
    dat$heidairy <- 10*(dat$dairyden/1.3)
    dat$heidairy <- ifelse(dat$heidairy > 10, 10, dat$heidairy)

    # heitotpro
    # total protein
    dat$totproden <- dat$lallmeat/(dat$TKCAL/1000)
    dat$heitotpro <- 5*(dat$totproden/2.5)
    dat$heitotpro <- ifelse(dat$heitotpro > 5, 5, dat$heitotpro)

    # heiseaplantpro
    # seaplant protein
    dat$seaplantden <- dat$lseaplant/(dat$TKCAL/1000)
    dat$heiseaplantpro <- 5*(dat$seaplantden/0.8)
    dat$heiseaplantpro <- ifelse(dat$heiseaplantpro > 5, 5, dat$heiseaplantpro)

    # heifattyacid
    # fatty acid
    dat$faratio <- ifelse(dat$TSFAT > 0,
                          dat$MONOPOLY / dat$TSFAT,
                          0)

    dat$heifattyacid <- ifelse(dat$TSFAT == 0 & dat$MONOPOLY == 0,
                               10,
                               NA)

    dat$heifattyacid <- ifelse(dat$TSFAT == 0 & dat$MONOPOLY > 0,
                               10,
                               dat$heifattyacid)

    dat$heifattyacid <- ifelse(dat$faratio >= 2.5,
                               10,
                               dat$heifattyacid)

    dat$heifattyacid <- ifelse(dat$faratio <= 1.2,
                               0,
                               dat$heifattyacid)

    dat$heifattyacid <- ifelse(is.na(dat$heifattyacid),
                               10*((dat$faratio-1.2)/(2.5-1.2)),
                               dat$heifattyacid)
    # heisodi
    # sodium
    dat$sodden <- dat$TSODI / dat$TKCAL

    dat$heisodi <- ifelse(dat$sodden <= 1.1,
                          10,
                          NA)
    dat$heisodi <- ifelse(dat$sodden >= 2.0,
                          0,
                          dat$heisodi)
    dat$heisodi <- ifelse(is.na(dat$heisodi),
                          10 - (10*(dat$sodden - 1.1)/(2.0-1.1)),
                          dat$heisodi)

    # heirefgrain
    # refined grain
    dat$refgrainnden <- dat$T_G_REFINED / (dat$TKCAL/1000)

    dat$heirefgrain <- ifelse(dat$refgrainnden <= 1.8,
                              10,
                              NA)
    dat$heirefgrain <- ifelse(dat$refgrainnden >= 4.3,
                              0,
                              dat$heirefgrain)

    dat$heirefgrain <- ifelse(is.na(dat$heirefgrain),
                              10 - (10*(dat$refgrainnden - 1.8)/(4.3-1.8)),
                              dat$heirefgrain)

    # heisofaas
    dat$sofa_perc <- 100* (dat$EMPTYCAL10/dat$TKCAL)

    dat$heisofaas <- ifelse(dat$sofa_perc >= 50,
                            0,
                            NA)

    dat$heisofaas <- ifelse(dat$sofa_perc <= 19,
                            20,
                            dat$heisofaas)

    dat$heisofaas <- ifelse(is.na(dat$heisofaas),
                           20 - (20*(dat$sofa_perc-19)/(50-19)),
                           dat$heisofaas)

    dat$HEI <- dat$heiveg + dat$heibngrn + dat$heitotfrt + dat$heiwholefrt + dat$heiwholegrain + dat$heidairy + dat$heitotpro + dat$heiseaplantpro + dat$heifattyacid + dat$heirefgrain + dat$heisofaas + dat$heisodi

    if (verbose) {

        dat

    }
    else {

        dat[,c("SEQN","RIDAGEYR","HEI")]

    }

}